python/mspasspy/db/database.py

```diff
@@ -574,45 +642,198 @@ def save_data(self, mspass_object, mode="promiscuous", storage_mode='gridfs', di
+                # before continuing.  pedantic kills data with any problems
+                # Cautious tries to fix the problem first
+                # Note many errors can be posted - one for each problem key-value pair
+                for k in copied_metadata:
+                    if save_schema.is_defined(k):
+                        if isinstance(copied_metadata[k], save_schema.type(k)):
+                            insertion_dict[k]=copied_metadata[k]
+                        else:
+                            if mode == 'pedantic':
+                                mspass_object.kill()
+                                message ='pedantic mode error:  key='+k
+                                if not save_schema.is_defined(k):
+                                    message += ' not defined in schema.'
+                                else:
+                                    value=copied_metadata[k]
+                                    message += ' type of stored value='+str(type(value))+' does not match schema expectation='+str(save_schema.type(k))
+                                mspass_object.elog.log_error('Database.save_data',
+                                    'message',ErrorSeverity.Invalid)
+                            else:
+                                # Careful if another mode is added here.  else means cautious in this logic
+                                 try:
+                                     # The following convert the actual value in a dict to a required type.
+                                     # This is because the return of type() is the class reference.
+                                     insertion_dict[k] = save_schema.type(k)(copied_metadata[k])
+                                 except Exception as err:
+                                     mspass_object.kill()
+                                     message = 'cautious mode error:  key=' + k
+                                     if save_schema.is_required(k):
+                                        message += ' Required key value could not be converted to required type='+str(save_schema.type(k))+' actual type='+str(type(copied_metadata[k]))
+                                     else:
+                                        message += ' Value has type='+str(type(copied_metadata[k])+' which cannot be converted to type='+str(save_schema.type(k)))
+                                     message += '\nPython error exception message caught:\n'
+                                     message += str(err)
+                                     mspass_object.elog.log_error('Database.save',
+                                           message,ErrorSeverity.Invalid)
+
+        # Note we jump here immediately if mspass_object was marked dead
+        # on entry.  Data can, however, be killed in metadata section
+        # above so we need repeat the test for live
+        wf_collection=self.db[wf_collection_name]
```

Comment:
After comparing the code in the master branch, I found the issue is here. Here we should not assign `self.db[wf_collection_name]` to the `wf_collection `. Instead, we should use `self[wf_collection_name]` because self itself is a Database object already! That's why we should directly access the collection name.

Reply:
I found update_metadata also has this problem. Anyway, now it works. But I am a bit confused why we could insert a document into `self.db[wf_collection_name]` because we didn't declare what self.db is.

Reply:
Bravo!   As I predicted it was a stupid little syntax error that was hard to spot.   I concur that I do not understand why the interpreter didn't flag that as an error, or why it didn't generate a runtime error.   So glad you found that.   

Can you push your changes?   Then I'll try to implement some of the changes we discussed yesterday in database.py.   Perhaps you can focus on modifying the test code to deal with the changes.  The biggest issue will be that most of the tests for save_data check the return.  If we convert that to returning the mspass_object all of them will have to change.

Reply:
PS.   Before you push your changes you might quickly change save_data and update_metadata to return the mspass_object.  That way you can be running the tests that use that change while I do the internal changes we discussed yesterday.   Changing the return should be pretty trivial.

Reply:
Sure, I will do it today.